### ISSUE TITLE
connection.close() results in disconnected event being fired instead of connection_close in case of open links

### ISSUE THREAD

**Issue by ramya-rao-a** [amqp/rhea#197]
@grs 

Calling `connection.close()` after closing all the open sender/receiver links, results in the `connection_close` event being fired. This is as expected. So far so good.

But, if we end up calling `connection.close()` without closing the sender/receiver links, the `disconnected` event is being fired.  Is this expected?

Can we not expect `connection.close()` to close all the sessions and links inside it and fire the `connection_close` event?

**Comment by grs** [comment]
> But, if we end up calling `connection.close()` without closing the sender/receiver links, the `disconnected` event is being fired. Neither `connection_close` nor `connection_error` events are fired. Is this expected?

No, I would expect to see a connection_close. Indeed if I alter simple_recv.py to not close the receiver link before closing the connection, I *do* see a connection_close event. Can you create a minimal example that shows the behaviour you describe?

**Comment by grs** [comment]
```
var container = require('rhea');

var args = require('./options.js').options({
    'h': { alias: 'host', default: 'localhost', describe: 'dns or ip name of server where you want to connect'},
    'p': { alias: 'port', default: 5672, describe: 'port to connect to'}
}).help('help').argv;

container.on('connection_close', function (context) {
    console.log('connection closed: %s', context.connection.options.container_id);
});

container.on('disconnected', function (context) {
    console.log('disconnected: %s', context.connection.options.container_id);
});

var c = container.connect({port: args.port, host: args.host, container_id:'A'});
c.open_receiver('foo');
c.open_receiver('bar').on('receiver_open', function(context) {
    context.connection.close();
});
container.connect({port: args.port, host: args.host, container_id:'B'}).open_receiver('baz');
```
If I understand correctly, this example does something similar. It has two connections and three receiver links open and then closes one of the connections. The connection_close is logged as expected, with no disconnected. If the server is then killed, the disconnected event is seen only for the second connection that was not closed.

**Comment by ramya-rao-a** [comment]
@grs I was not able to create a sample that directly uses `rhea`, but I was able to get a sample using [rhea-promise](https://github.com/amqp/rhea-promise)

Use `DEBUG=rhea-promise*` and in the below example, add in details for host/port etc

```javascript
const { Connection, delay } = require("rhea-promise");

const host = "";
const username = "";
const password = "";
const port = "";
const receiverAddress = "";

async function main(){
  const connection = new Connection({
    transport: "tls",
    host: host,
    hostname: host,
    username: username,
    password: password,
    port: port,
    reconnect: false
  });
  const receiverOptions = {
    name: "receiver-1",
    source: {
      address: receiverAddress
    }
  };

  await connection.open();
  const receiver = await connection.createReceiver(receiverOptions);
  // sleeping for 1 seconds to let the receiver receive messages
  await delay(1000);
  // await receiver.close();

  console.log("Attempting to close the connection");
  await connection.close();
}

main().catch(err => console.log(err));
```

With the line `await receiver.close();` uncommented, we get the `connection_close` event.

![image](https://user-images.githubusercontent.com/16890566/53706238-0ce6d800-3dde-11e9-840d-7335fd845a31.png)


With the same commented, I get the `disconnected` event. 60 seconds later, we get the timeout error from https://github.com/amqp/rhea-promise/blob/master/lib/connection.ts#L310



![image](https://user-images.githubusercontent.com/16890566/53706194-e759ce80-3ddd-11e9-8ff9-0f63a443e47a.png)


Interestingly I cannot get a repro of this on my Mac, I see it only on Windws

**Comment by grs** [comment]
I am unable to reproduce (on linux). What do you see if you set DEBUG=* on the platform where you see the problem?
